Lib/asyncio/base_events.py

```diff
@@ -287,32 +305,47 @@ def __init__(self, loop, sockets, protocol_factory, ssl_context, backlog,
         self._ssl_context = ssl_context
         self._ssl_handshake_timeout = ssl_handshake_timeout
         self._ssl_shutdown_timeout = ssl_shutdown_timeout
-        self._serving = False
+        self._state = _ServerState.INITIALIZED
         self._serving_forever_fut = None
 
     def __repr__(self):
         return f'<{self.__class__.__name__} sockets={self.sockets!r}>'
 
     def _attach(self, transport):
-        assert self._sockets is not None
+        if self._state != _ServerState.SERVING:
+            raise RuntimeError("server is not serving, cannot attach transport")
         self._clients.add(transport)
 
     def _detach(self, transport):
         self._clients.discard(transport)
-        if len(self._clients) == 0 and self._sockets is None:
+        if self._state == _ServerState.CLOSED and len(self._clients) == 0:
             self._wakeup()
 
     def _wakeup(self):
+        if self._state == _ServerState.CLOSED:
+            self._state = _ServerState.SHUTDOWN
+        elif self._state == _ServerState.SHUTDOWN:
+            # gh109564: the wakeup method has two possible call-sites,
+            # through an explicit call Server.close(), or indirectly through
+            # Server._detach() by the last connected client.
+            return
+        else:
+            raise RuntimeError(f"server {self!r} can only wakeup waiters after closing")
+
         waiters = self._waiters
         self._waiters = None
         for waiter in waiters:
             if not waiter.done():
                 waiter.set_result(None)
 
     def _start_serving(self):
-        if self._serving:
+        if self._state == _ServerState.INITIALIZED:
+            self._state = _ServerState.SERVING
+        elif self._state == _ServerState.SERVING:
             return
-        self._serving = True
+        else:
+            raise RuntimeError(f'server {self!r} is closed')
```

Comment:
I'm concerned about this one too. The user definitely knows that the server is closed, that's why they're trying to start it :wink:

Instead, we probably should say that the server was *already* started and then closed, so something like this:
```suggestion
            raise RuntimeError(f'server {self!r} was already started and then closed')
```